src/Features/CSharp/Portable/Snippets/ConsoleSnippetProvider.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Microsoft.CodeAnalysis.Snippets;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Completion.CompletionProviders.Snippets
+{
+    [Export(typeof(ISnippetProvider)), Shared]
+    internal class ConsoleSnippetProvider : AbstractSnippetProvider
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public ConsoleSnippetProvider()
+        {
+        }
+
+        protected override async Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
+            {
+                return false;
+            }
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var syntaxContext = (CSharpSyntaxContext)document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(document, semanticModel, position, cancellationToken);
+
+            if (!ShouldDisplaySnippet(syntaxContext))
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static bool ShouldDisplaySnippet(CSharpSyntaxContext context)
+        {
+            var token = context.LeftToken;
+            var isDirectlyInUndesirableLocation = token.GetAncestors<SyntaxNode>()
+                .Any(node => node.IsKind(SyntaxKind.ParameterList) ||
+                             node.IsKind(SyntaxKind.SimpleLambdaExpression) ||
+                             node.IsKind(SyntaxKind.ArgumentList) ||
+                             node.IsKind(SyntaxKind.RecordDeclaration) ||
+                             node.IsKind(SyntaxKind.ObjectCreationExpression) ||
+                             node.IsKind(SyntaxKind.SwitchExpression));
```

Comment:
how was this list created?

shouldn't we jsut offer this in any statement context?